### PR TITLE
Fix realloc method for ChibiOS

### PR DESCRIPTION
**Pull request by bugobliterator**
realloc method is left undefined, hence calling libc method that breaks realloc in AP_ExpandingArray used in Dijkstra Avoidance code.

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/Util.cpp [hunk 1]
```diff
@@ -86,6 +86,26 @@ void *Util::allocate_heap_memory(size_t size)
...
+  realloc implementation thanks to wolfssl, used by AP_Scripting
+ */
+void *Util::std_realloc(void *addr, size_t size)
+{
+    if (size == 0) {
+       free(addr);
+       return nullptr;
+    }
+    if (addr == nullptr) {
+        return malloc(size);
```
Comment (khancyr) [inline libraries/AP_HAL_ChibiOS/Util.cpp]: Didn't we prefere calloc?

Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Util.cpp]: malloc in ArduPilot does zero fill, so equivalent to calloc

**Comment by tridge** [comment]
@bugobliterator I presume the breakage is because it doesn't zero-fill the added memory?
If we are going to ban ::realloc() then we should ensure we get a build error if anyone uses it. Also, why the std_realloc() name? Maybe it should be realloc_type() and use the type like malloc_type() does?

**Comment by bugobliterator** [comment]
> @bugobliterator I presume the breakage is because it doesn't zero-fill the added memory?
> If we are going to ban ::realloc() then we should ensure we get a build error if anyone uses it. Also, why the std_realloc() name? Maybe it should be realloc_type() and use the type like malloc_type() does?

It just failed to realloc, my assumption is that we are using malloc that we defined and trying to call libc realloc on that pointer, I don't think that would work.

Regarding the name, I don't care much about it, I just wanted some way to distance from heap_realloc allocator from specified heap with common heap realloc. So whatever your preference I can implement that.

**Comment by bugobliterator** [comment]
@tridge Also introduced error when libc realloc is used for ChibiOS builds.

**Comment by bugobliterator** [comment]
@tridge @WickedShell the check added above to produce error whenever realloc is being used, has detected use of realloc under lua scripting, can you review and push the necessary fix to this branch?

**Comment by WickedShell** [comment]
@bugobliterator I can't push to it, but you can just comment out that whole section. It's actually unused, as we don't ever call that function.

``` diff
diff --git a/libraries/AP_Scripting/lua/src/lauxlib.c b/libraries/AP_Scripting/lua/src/lauxlib.c
index 8bdada50a7..b4948e3395 100644
--- a/libraries/AP_Scripting/lua/src/lauxlib.c
+++ b/libraries/AP_Scripting/lua/src/lauxlib.c
@@ -1005,29 +1005,29 @@ LUALIB_API const char *luaL_gsub (lua_State *L, const char *s, const char *p,
 }
 
 
-static void *l_alloc (void *ud, void *ptr, size_t osize, size_t nsize) {
-  (void)ud; (void)osize;  /* not used */
-  if (nsize == 0) {
-    free(ptr);
-    return NULL;
-  }
-  else
-    return realloc(ptr, nsize);
-}
-
-
-static int panic (lua_State *L) {
-  lua_writestringerror("PANIC: unprotected error in call to Lua API (%s)\n",
-                        lua_tostring(L, -1));
-  return 0;  /* return to Lua to abort */
-}
-
-
-LUALIB_API lua_State *luaL_newstate (void) {
-  lua_State *L = lua_newstate(l_alloc, NULL);
-  if (L) lua_atpanic(L, &panic);
-  return L;
-}
+//static void *l_alloc (void *ud, void *ptr, size_t osize, size_t nsize) {
+//  (void)ud; (void)osize;  /* not used */
+//  if (nsize == 0) {
+//    free(ptr);
+//    return NULL;
+//  }
+//  else
+//    return realloc(ptr, nsize);
+//}
+//
+//
+//static int panic (lua_State *L) {
+//  lua_writestringerror("PANIC: unprotected error in call to Lua API (%s)\n",
+//                        lua_tostring(L, -1));
+//  return 0;  /* return to Lua to abort */
+//}
+//
+//
+//LUALIB_API lua_State *luaL_newstate (void) {
+//  lua_State *L = lua_newstate(l_alloc, NULL);
+//  if (L) lua_atpanic(L, &panic);
+//  return L;
+//}
 
 
 LUALIB_API void luaL_checkversion_ (lua_State *L, lua_Number ver, size_t sz) {
```

**Comment by rmackay9** [comment]
This is included in Copter-4.0.4-rc1

**Pull request merged**